### PR TITLE
⚠️ Enforce version skew policy in rollout undo

### LINKED ISSUE #8408: Prevent k8s version skew policy violations
### What would you like to be added (User Story)?

As an operator I'd like to validate the Kubernetes version skew policy when I run `clusterctl alpha rollout undo`.

### Detailed Description

The existing clusterclt rollout undo command might violate the [version skew policy](https://kubernetes.io/releases/version-skew-policy/) easily and accidentally. For example, if a user has MachineSet (revision=1, k8s version=1.22) and MachineSet (revision=4, k8s version=1.25) and mistakenly triggers rollback from revision 4 to 1, there's no way to stop it. I'd like to suggest that clusterctl raises an error if users break the version skew policy and allow the operation only if a user sets --force option. 

### Anything else you would like to add?

Possible Concerns and Solutions:
- This feature changes the default behavior of `rollout` and makes negative impact on the existing end users.
  - -> violation of version skew policy is a more serious problem. As an alternative, we can skip validation when the `version` field is empty.
- Isn't interactive-mode better than `--force` option?
  - -> users who run this command from scripts definitely need `--force` options rather than interactive-mode.
- Which control plane providers are supported?
  - -> KubeadmControlPlane, but is extensible.

Related issue: #3439 

### Label(s) to be applied

/kind feature
/area clusterctl

### PR DESCRIPTION
**Pull request by hiromu-a5a**
**What this PR does / why we need it**:

What this PR does:

Changed `clusterctl rollback undo` to raise an error if users break the version skew policy and allow such violation only if a user sets --force option. It is noted that this change may affect user experiences as the behavior of `rollout undo` will be changed.

why we need it: 

The existing clusterclt rollout undo command might violate the [version skew policy](https://kubernetes.io/releases/version-skew-policy/) easily and accidentally. For example, if a user has MachineSet (revision=1, k8s version=1.22) and MachineSet (revision=4, k8s version=1.25) and mistakenly triggers rollback from revision 4 to 1, there's no way to stop it. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8408

### CODE REVIEW

File: cmd/clusterctl/client/alpha/rollout_rollbacker.go [hunk 1]
```diff
@@ -78,3 +120,11 @@ func rollbackMachineDeployment(proxy cluster.Proxy, md *clusterv1.MachineDeploym
 	md.Spec.Template = revMSTemplate
 	return patchHelper.Patch(ctx, md)
 }
+
+func validateVersionSkewPolicy(cpVersion semver.Version, msVersion semver.Version) error {
+	if cpVersion.Major != msVersion.Major || cpVersion.Minor < msVersion.Minor || cpVersion.Minor-msVersion.Minor > 2 {
+		return errors.Errorf("version skew between ControlPlane %v and MachineSet %v is not supporeted by Kubernetes version skew policy."+
+			" If you want to rollback anyway, use --force option.", cpVersion, msVersion)
```
Comment (vincepri) [inline cmd/clusterctl/client/alpha/rollout_rollbacker.go:127]: Wouldn't the validation webhook trigger an error in this case?

Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/rollout_rollbacker.go]: Do you mean [this](https://github.com/hiromu-a5a/cluster-api/blob/c1c328f9edc610ee519f7ea3c8e0318acdc16960/internal/controllers/machineset/machineset_preflight.go#L173-L178) ?
Strangely, I can skip this validation when I run `clusterctl alpha rollout undo`. Even I tried to rollback by modifying MS directly, it still works. Maybe I missed something.

File: cmd/clusterctl/client/alpha/rollout_rollbacker.go [hunk 2]
```diff
@@ -78,3 +120,11 @@ func rollbackMachineDeployment(proxy cluster.Proxy, md *clusterv1.MachineDeploym
 	md.Spec.Template = revMSTemplate
 	return patchHelper.Patch(ctx, md)
 }
+
+func validateVersionSkewPolicy(cpVersion semver.Version, msVersion semver.Version) error {
+	if cpVersion.Major != msVersion.Major || cpVersion.Minor < msVersion.Minor || cpVersion.Minor-msVersion.Minor > 2 {
+		return errors.Errorf("version skew between ControlPlane %v and MachineSet %v is not supporeted by Kubernetes version skew policy."+
```
[the remark this answers is not in the document]
Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/rollout_rollbacker.go]: I've fixed in the latest commit. Sorry for the delay.

**Comment by Jont828** [comment]
@hiromu-a5a Looks like you need to rebase. Happy to give it a review once it's up to date.

**Comment by hiromu-a5a** [comment]
@Jont828 Thank you for reminding me and I'm sorry for the delayed response. I've rebased it. Could you give me a review?

File: cmd/clusterctl/client/alpha/cluster.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+)
+
+// getCluster retrieves the Cluster object corresponding to the name and namespace specified.
+func getCluster(ctx context.Context, proxy cluster.Proxy, name, namespace string) (*clusterv1.Cluster, error) {
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/cluster.go:29]: Is it worth moving this to a util file or the util package? I'd imagine there are lots of instances where we would want to get a cluster outside of clusterctl.

File: cmd/clusterctl/client/alpha/rollout_rollbacker.go [hunk 4]
```diff
@@ -49,7 +52,7 @@ func (r *rollout) ObjectRollbacker(ctx context.Context, proxy cluster.Proxy, ref
 }
 
 // rollbackMachineDeployment will rollback to a previous MachineSet revision used by this MachineDeployment.
-func rollbackMachineDeployment(ctx context.Context, proxy cluster.Proxy, md *clusterv1.MachineDeployment, toRevision int64) error {
+func rollbackMachineDeployment(ctx context.Context, proxy cluster.Proxy, md *clusterv1.MachineDeployment, toRevision int64, force bool) error {
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_rollbacker.go:55]:
```suggestion
func rollbackMachineDeployment(ctx context.Context, proxy cluster.Proxy, md *clusterv1.MachineDeployment, toRevision int64, forceRollback bool) error {
```
nit: for extra clarity

File: cmd/clusterctl/cmd/rollout/undo.go [hunk 5]
```diff
@@ -66,6 +67,7 @@ func NewCmdRolloutUndo(cfgFile string) *cobra.Command {
 		"Context to be used within the kubeconfig file. If empty, current context will be used.")
 	cmd.Flags().StringVarP(&undoOpt.namespace, "namespace", "n", "", "Namespace where the resource(s) reside. If unspecified, the defult namespace will be used.")
 	cmd.Flags().Int64Var(&undoOpt.toRevision, "to-revision", undoOpt.toRevision, "The revision to rollback to. Default to 0 (last revision).")
+	cmd.Flags().BoolVarP(&undoOpt.force, "force", "f", undoOpt.force, "Force rollback without validating the Kubernetes version skew policy.")
```
Comment (Jont828) [inline cmd/clusterctl/cmd/rollout/undo.go:70]: Just to be clear, the rename was only for the func parameter, I like the `--force` flag as it is.

**Review by Jont828 — COMMENTED** [review]
Seems fine to me except for a couple tiny things!

**Comment by hiromu-a5a** [comment]
Passed `pull-cluster-api-test-main` in my environment. Try again.

File: cmd/clusterctl/client/alpha/cluster.go (hunk 3 above)
Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/cluster.go]: I removed `alpha/cluster.go` and changed to use `util.GetClusterByName` instead. I forgot why I re-implement `getCluster` at that moment I created this PR, but it still passed e2e-test after replacing it with `GetClusterByName.

**Comment by hiromu-a5a** [comment]
@Jont828 Sorry for the late reply. I've applied your comments. Thank you for the helpful comments.

File: cmd/clusterctl/client/rollout.go [hunk 6]
```diff
@@ -84,6 +84,9 @@ type RolloutUndoOptions struct {
 
 	// Revision number to rollback to when issuing the undo command.
 	ToRevision int64
+
+	// Force execute rollback
```
Comment (Jont828) [inline cmd/clusterctl/client/rollout.go:88]:
```suggestion
	// Force execute rollback.
```

**Review by Jont828 — COMMENTED** [review]
LGTM pending a nit. I'd like to get a second pair of eyes on this if anyone else has the bandwith, though.

**Comment by Jont828** [comment]
@hiromu-a5a Are you still working on this? If not we might want to close this PR as it's been without activity for almost 6 months like other PRs.

**Comment by sbueringer** [comment]
At this point it's very very likely that we will deprecate and remove the revision management, see https://github.com/kubernetes-sigs/cluster-api/issues/10479

**Comment by hiromu-a5a** [comment]
I am really sorry for the late reply. We are shifting to GitOps and lost motivation to clusterctl.

**Pull request closed without merging**